skbio/metadata/_interval.py

```diff
@@ -0,0 +1,599 @@
+            objects. An satisfying interval feature only need to overlap with
+            one location. Default (``None``) means all ``Interval``s meet
+            this requirement.
+
+        metadata : dict
+            A dictionary of key word attributes associated with the
+            ``Interval`` object. It specifies what metadata keywords and
+            values to look for. Default (``None``) means all ``Interval``s
+            meet this requirement.
+
+        Returns
+        -------
+        generator, Interval
+            A generator of Interval objects satisfying the search criteria.
+        """
+        if locations is None:
+            for intvl in self._query_attribute(metadata):
+                yield intvl
+        else:
+            for loc in locations:
+                intvls = self._query_interval(loc)
+                if metadata is None:
+                    metadata = {}
+                for intvl in self._query_attribute(metadata, intvls):
+                    yield intvl
+
+    @experimental(as_of='0.4.2-dev')
+    def drop(self, locations=None, metadata=None):
+        """Drops Interval objects according to a specified query.
+
+        Locations are first queried from the IntervalMetadata object
+        using the query functionality. These locations are then dropped
+        from the ``IntervalTree`` object before being deleted.
+
+        Parameters
+        ----------
+        locations : iterable of tuple of ints
+            A list of locations associated with the Interval object.
+        metadata : dict
+            A dictionary of key word attributes associated with the
+            Interval object.
+
+        Examples
+        --------
+        >>> from skbio.metadata import IntervalMetadata
+        >>> im = IntervalMetadata()
+        >>> im.add(locations=[(0, 2), (4, 7)], metadata={'name': 'sagA'})   # doctest: +ELLIPSIS
+        Interval(interval_metadata=..., locations=[(0, 2), (4, 7)], boundaries=[(True, True), (True, True)], metadata={'name': 'sagA'})
+        >>> im.add(locations=[(40, 70)], metadata={'name': 'sagB'})   # doctest: +ELLIPSIS
+        Interval(interval_metadata=..., locations=[(40, 70)], boundaries=[(True, True)], metadata={'name': 'sagB'})
+        >>> im.drop(metadata={'name': 'sagA'})
+        >>> im   # doctest: +ELLIPSIS
+        1 interval features
+        -------------------
+        Interval(interval_metadata=..., locations=[(40, 70)], boundaries=[(True, True)], metadata={'name': 'sagB'})
+        """
+        to_delete = {id(f) for f in
+                     self.query(locations=locations,
+                                metadata=metadata)}
+
+        new_intvls = []
+        # iterate through queries and drop them
+        for intvl in self._intervals:
+            if id(intvl) not in to_delete:
+                new_intvls.append(intvl)
+
+        self._intervals = new_intvls
+        self._is_stale_tree = True
```

Comment:
we want to defer the rebuilding as much as possible, right?

I like you the other idea and made a decorator to all the methods of the class, less confusion for future developer. Thanks!

I also addressed your other comments. good for another round of review, @wasade